doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,1693 @@
+avoided.
+
+- **Don't just return an expression.**
+
+  In the above example, we might have been tempted to write
+
+  ```py
+  >>> from sympy import cos
+  >>> class versin(Function):
+  ...     @classmethod
+  ...     def eval(cls, x):
+  ...         # !! Not actually a good eval() method !!
+  ...         return 1 - cos(x)
+  ```
+
+  However, this would make it so that `versin(x)` would *always* return `1 -
+  cos(x)`, regardless of what `x` is. If all you want is a quick shorthand to
+  `1 - cos(x)`, that is fine, but would be much simpler and more explicit to
+  just [use a Python function as described
+  above](custom-functions-fully-evaluated). If we defined `versin` like this,
+  it would never actually be represented as `versin(x)`, and none of the other
+  behavior we define below would matter, because the other behaviors we are
+  going to define on the `versin` class only apply when the returned object is
+  actually a `versin` instance. So for example, `versin(x).diff(x)` would
+  actually just be `(1 - cos(x)).diff(x)`, instead of calling [the `fdiff()`
+  method we define below](custom-functions-differentiation).
+
+  ```{admonition} Key Point
+
+  **The purpose of `eval()` is not to define what the function *is*,
+  mathematically, but rather to specify on what inputs it should automatically
+  evaluate.** The mathematical definition of a function is determined through
+  the specification of various mathematical properties with the methods
+  outlined below, like [numerical evaluation](custom-functions-evalf),
+  [differentiation](custom-functions-differentiation), and so on.
+
+  ```
+
+  If you find yourself doing this, you should think about what you actually
+  want to achieve. If you just want a shorthand function for an expression, it
+  will be simpler to just [define a Python
+  function](custom-functions-fully-evaluated). If you really do want a
+  symbolic function, think about when you want it to evaluate to something
+  else and when you want it to stay unevaluated. One option is to make your
+  function unevaluated in `eval()` and define a [`doit()`
+  method](custom-functions-doit) to evaluate it.
+
+- **Avoid too much automatic evaluation.**
+
+  It is recommended to minimize what is evaluated automatically by `eval()`.
+  It is typically better to put more advanced simplifications in [other
+  methods](custom-functions-rewriting-and-simplification), like
+  [`doit()`](custom-functions-doit). Remember that whatever you define for
+  automatic evaluation will *always* evaluate.[^evaluate-footnote] As in the
+  previous point, if you evaluate every value, there is little point to even
+  having a symbolic function in the first place. For example, we might be
+  tempted to evaluate some trig identities on `versin` in `eval()`, but then
+  these identities would always evaluate, and it wouldn't be possible to
+  represent one half of the identity.
```

Comment:
It might be good to have an example here showing what the implications would be and why we might not want to use automatic evaluation.

Reply:
I added an example. See what you think.